File: bumperbot_cpp_examples/src/simple_qos_subscriber.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

using std::placeholders::_1;

class SimpleQoSSubscriber: public rclcpp::Node
{
    public:
        SimpleQoSSubscriber(): Node("simple_subscriber"), qos_profile_sub_(10)
        {

            declare_parameter<std::string>("reliability", "system_default");
            declare_parameter<std::string>("durability", "system_default");

            const auto reliability = get_parameter("reliability").as_string();
            const auto durability = get_parameter("durability").as_string();

            if(reliability == "best_effort"){
                qos_profile_sub_.reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT);
                RCLCPP_INFO(get_logger(), "[Reliability]: Best Effort");
            }else if(reliability == "reliable"){
                qos_profile_sub_.reliability(RMW_QOS_POLICY_RELIABILITY_RELIABLE);
                RCLCPP_INFO(get_logger(), "[Reliability]: Reliable");
            }else if(reliability == "system_default"){
                qos_profile_sub_.reliability(RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT);
                RCLCPP_INFO(get_logger(), "[Reliability]: System Default");
            }else{
                RCLCPP_ERROR(get_logger(), "Selected Reliability QoS doesn't exists");
            }

            if(durability == "transient_local"){
                qos_profile_sub_.durability(RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL);
                RCLCPP_INFO(get_logger(), "[Reliability]: transient local");
            }else if(durability == "volatile"){
                qos_profile_sub_.durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);
                RCLCPP_INFO(get_logger(), "[Reliability]: Volatile");
            }else if(durability == "system_default"){
                qos_profile_sub_.durability(RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT);
                RCLCPP_INFO(get_logger(), "[Reliability]: System Default");
            }else{
                RCLCPP_ERROR(get_logger(), "Selected Reliability QoS doesn't exists");
            }
            sub_ = create_subscription<std_msgs::msg::String>("chatter", 
            qos_profile_sub_, 
            std::bind(&SimpleQoSSubscriber::msgCallback, this, _1));
        }

    private:
        rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
        rclcpp::QoS qos_profile_sub_;

        
        void msgCallback(const std_msgs::msg::String &msg) const
        {
            
            RCLCPP_INFO_STREAM(get_logger(), "counter " << msg.data.c_str());
        }
};

int main(int argc, char *argv[]){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleQoSSubscriber>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: bumperbot_python_examples/bumperbot_python_examples/simple_qos_subscriber.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSDurabilityPolicy
from std_msgs.msg import String

class SimpleQoSSubscriber(Node):
    def __init__(self):
        super().__init__("simple_qos_subscriber")

        self.qos_profile_ = QoSProfile(depth=10)

        self.declare_parameter("reliability", "system_default")
        self.declare_parameter("durability", "system_default")

        reliability = self.get_parameter("reliability").get_parameter_value().string_value
        durability = self.get_parameter("durability").get_parameter_value().string_value

        if reliability == "best_effort":
            self.qos_profile_.reliability = QoSReliabilityPolicy.BEST_EFFORT
            self.get_logger().info("[Reliability] : Best Effort")
        elif reliability == "reliable":
            self.qos_profile_.reliability = QoSReliabilityPolicy.RELIABLE
            self.get_logger().info("[Reliability] : Reliable")
        elif reliability == "system_default":
            self.qos_profile_.reliability = QoSReliabilityPolicy.SYSTEM_DEFAULT
            self.get_logger().info("[Reliability] : System Default")
        else:
            self.get_logger().error(f"Selected Reliability QoS: {reliability} doesn't exists")
            return

        if durability == "transient_local":
            self.qos_profile_.durability = QoSDurabilityPolicy.TRANSIENT_LOCAL
            self.get_logger().info("[Durability] : Transient Local")
        elif durability == "volatile":
            self.qos_profile_.durability = QoSDurabilityPolicy.VOLATILE
            self.get_logger().info("[Durability] : Volatile")
        elif durability == "system_default":
            self.qos_profile_.reliability = QoSDurabilityPolicy.SYSTEM_DEFAULT
            self.get_logger().info("[Durability] : System Default")
        else:
            self.get_logger().error(f"Selected Reliability QoS: {durability} doesn't exists")
            return


        self.sub_ = self.create_subscription(String, "chatter", self.subscriberCallback, self.qos_profile_)

    def subscriberCallback(self, msg):
        self.get_logger().info(f"counter {msg.data}")


def main():
    rclpy.init()
    node = SimpleQoSSubscriber()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()